services/venice-controller/src/test/java/com/linkedin/venice/controller/TestProtocolVersionAutoDetectionService.java

```diff
@@ -0,0 +1,240 @@
+  }
+
+  @Test
+  public void testProtocolVersionDetectionWithNoUpdate() throws Exception {
+    // When version is -1, no need to update
+    Map<String, Long> adminTopicMetadataMap = AdminTopicMetadataAccessor
+        .generateMetadataMap(Optional.of(1L), Optional.of(-1L), Optional.of(1L), Optional.of(-1L));
+
+    doReturn(adminTopicMetadataMap).when(admin).getAdminTopicMetadata(clusterName, Optional.empty());
+
+    ProtocolVersionAutoDetectionService localProtocolVersionAutoDetectionService =
+        new ProtocolVersionAutoDetectionService(
+            clusterName,
+            admin,
+            mock(ProtocolVersionAutoDetectionStats.class),
+            DEFAULT_SLEEP_INTERVAL_MS);
+
+    localProtocolVersionAutoDetectionService.startInner();
+
+    TestUtils.waitForNonDeterministicAssertion(5, TimeUnit.SECONDS, () -> {
+      verify(admin, atLeastOnce()).getAdminOperationVersionFromControllers(clusterName);
+      verify(adminConsumerService, never()).updateAdminOperationProtocolVersion(anyString(), anyLong());
+    });
+
+    localProtocolVersionAutoDetectionService.stopInner();
+  }
+
+  @Test
+  public void testGetLocalAdminOperationProtocolVersionForAllConsumers() {
+    ProtocolVersionAutoDetectionService localProtocolVersionAutoDetectionService =
+        new ProtocolVersionAutoDetectionService(
+            clusterName,
+            admin,
+            mock(ProtocolVersionAutoDetectionStats.class),
+            DEFAULT_SLEEP_INTERVAL_MS);
+    long smallestVersion = localProtocolVersionAutoDetectionService
+        .getSmallestLocalAdminOperationProtocolVersionForAllConsumers(clusterName);
+
+    assertEquals(smallestVersion, 1L);
+  }
+
+  @Test
+  public void testGetLocalAdminOperationVersionWithFailRequest() {
+    AdminOperationProtocolVersionControllerResponse failResponse =
+        new AdminOperationProtocolVersionControllerResponse();
+    failResponse.setError("Fail to connect to the controller");
+
+    Map<String, ControllerClient> controllerClientMap = new HashMap<>();
+    ControllerClient mockControllerClient = mock(ControllerClient.class);
+    controllerClientMap.put(REGION_DC_1, mockControllerClient);
+    controllerClientMap.put(REGION_DC_2, mockControllerClient);
+    doReturn(controllerClientMap).when(admin).getControllerClientMap(CLUSTER_VENICE_0);
+    when(mockControllerClient.getAdminOperationProtocolVersionFromControllers(CLUSTER_VENICE_0))
+        .thenReturn(getAdminOperationProtocolVersionResponse(REGION_DC_1, CLUSTER_VENICE_0))
+        .thenReturn(failResponse);
+
+    try {
+      ProtocolVersionAutoDetectionService localProtocolVersionAutoDetectionService =
+          new ProtocolVersionAutoDetectionService(
+              clusterName,
+              admin,
+              mock(ProtocolVersionAutoDetectionStats.class),
+              DEFAULT_SLEEP_INTERVAL_MS);
+
+      localProtocolVersionAutoDetectionService
+          .getSmallestLocalAdminOperationProtocolVersionForAllConsumers(clusterName);
```

Comment:
We probably need to add another fail() assertion after this line, if we expect this line to fail.